File: leetcode/0019. Remove Nth Node From End of List/Solution.swift
// https://leetcode.com/problems/remove-nth-node-from-end-of-list

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

class Solution {
    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        var p1 = head
        var p2 = head

        for _ in 0..<n {
            p1 = p1?.next
        }
        
        var prev: ListNode? = nil
        while p1 != nil {
            p1 = p1?.next
            prev = p2
            p2 = p2?.next
        }

        var head = head
        if prev != nil {
            prev?.next = p2?.next
        } else {
            head = head?.next
        }
        
        return head
    }
}

File: leetcode/0019. Remove Nth Node From End of List/Solution.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        
        fast_p = slow_p = dummy
        for i in range(1, n + 2):
            fast_p = fast_p.next
            
        while fast_p:
            fast_p = fast_p.next
            slow_p = slow_p.next
            
        slow_p.next = slow_p.next.next
        
        return dummy.next